x/subscription/types/calendar.go

```diff
@@ -0,0 +1,172 @@
+package types
+
+// port from https://github.com/esmil/musl/blob/master/src/time/__secs_to_tm.c
+
+const (
+	leaPoch        = 946684800 + 86400*(31+29) // 2000-03-01
+	daysPer400Y    = 365*400 + 97
+	daysPer100Y    = 365*100 + 24
+	daysPer4Y      = 365*4 + 1
+	daysBefore1970 = 719162
+)
+
+var (
+	// start at March
+	daysInMonth = [...]int{31, 30, 31, 30, 31, 31, 30, 31, 30, 31, 31, 29}
+	// start at Jan
+	daysInMonthNoLeap = [...]int{31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31}
+	daysInMonthLeap   = [...]int{31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31}
+	// start at January
+	daysBeforeMonth = [...]int{0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334}
+	// DaysInMonthSet start at Jan
+	DaysInMonthSet []BitSet
+)
+
+func init() {
+	for _, days := range daysInMonthLeap {
+		bs := NewBitSet()
+		for i := 1; i <= days; i++ {
+			err := bs.Set(uint(i))
+			if err != nil {
+				panic(err)
+			}
+		}
+		DaysInMonthSet = append(DaysInMonthSet, bs)
+	}
+}
+
+// TimeStruct is struct tm
+type TimeStruct struct {
+	Year   int
+	Month  int // 1 - 12
+	Mday   int // 1 -
+	Wday   int // 0 - 6, start from Sunday
+	Yday   int // 1 -
+	Hour   int // 0 - 23
+	Minute int // 0 - 59
+	Second int // 0 - 59
+}
+
+// Timestamp calculate utc timestamp from time struct
+func (tm TimeStruct) Timestamp() int64 {
+	days := int64(DateToOrdinal(tm.Year, tm.Month, tm.Mday) - daysBefore1970)
+	return days*86400 + int64(tm.Hour)*3600 + int64(tm.Minute)*60 + int64(tm.Second)
+}
+
+// SecsToTM converts timestamp to TimeStruct
+func SecsToTM(t int64, tzoffset uint32) TimeStruct {
+	// add timezone offset
+	t += int64(tzoffset)
+
+	secs := t - leaPoch
+	days, remsecs := divmod(secs, 86400)
+
+	_, wDay := divmod(3+days, 7)
+
+	qcCycles, remDays := divmod(days, daysPer400Y)
+
+	cCycles := remDays / daysPer100Y
+	if cCycles == 4 {
+		cCycles--
+	}
+	remDays -= cCycles * daysPer100Y
+
+	qCycles := remDays / daysPer4Y
+	if qCycles == 25 {
+		// TODO I think this path is never reached
```

Comment:
maybe year 2074 or something like that?

Reply:
I tried this one, it don't reach here, I've tried lots of cases. 😄